src/ql/pass/map/qubits/map/detail/alter.cc

```diff
@@ -1,264 +1,89 @@
-        QL_FATAL("Mapper option maxfidelity has been disabled");
-        // score = quick_fidelity(past.lg);
-    } else {
-        score = past.get_max_free_cycle() - base_past.get_max_free_cycle();
-    }
-    score_valid = true;
+void Alter::extend(Past curr_past, const Past &base_past) {
+    QL_ASSERT(!score_valid && "Alter::extend() can only be called once!");
+    add_swaps(curr_past);
+    set_score(curr_past.get_max_free_cycle() - base_past.get_max_free_cycle());
 }
 
-/**
- * Split the path. Starting from the representation in the total attribute,
- * generate all split path variations where each path is split once at any
- * hop in it. The intention is that the mapped two-qubit gate can be placed
- * at the position of that hop. All result paths are added/appended to the
- * given result list.
- *
- * When at the hop of a split a two-qubit gate cannot be placed, the split
- * is not done there. This means at the end that, when all hops are
- * inter-core, no split is added to the result.
- *
- * distance=5   means length=6  means 4 swaps + 1 CZ gate, e.g.
- * index in total:      0           1           2           length-3        length-2        length-1
- * qubit:               2   ->      5   ->      7   ->      3       ->      1       CZ      4
- */
-void Alter::split(utils::List<Alter> &result) const {
-    // QL_DOUT("Split ...");
-
-    utils::UInt length = total.size();
-    QL_ASSERT (length >= 2);   // distance >= 1 so path at least: source -> target
-    for (utils::UInt rightopi = length - 1; rightopi >= 1; rightopi--) {
-        utils::UInt leftopi = rightopi - 1;
-        QL_ASSERT (leftopi >= 0);
-        // QL_DOUT("... leftopi=" << leftopi);
-        // leftopi is the index in total that holds the qubit that becomes the left operand of the gate
-        // rightopi is the index in total that holds the qubit that becomes the right operand of the gate
-        // rightopi == leftopi + 1
-        if (platform->topology->is_inter_core_hop(total[leftopi], total[rightopi])) {
-            // an inter-core hop cannot execute a two-qubit gate, so is not a valid alternative
-            // QL_DOUT("... skip inter-core hop from qubit=" << total[leftopi] << " to qubit=" << total[rightopi]);
-            continue;
+utils::List<Alter> Alter::create_from_path(const ir::PlatformRef &platform, const ir::BlockBaseRef &block, const OptionsRef &options, ir::CustomInstructionRef gate, std::list<utils::UInt> path) {
+    utils::List<Alter> result;
+
+    std::shared_ptr<std::list<utils::UInt>> shared_path(new std::list<utils::UInt>(path));
+
+    QL_ASSERT (shared_path->size() >= 2);
+    
+    auto leftOpIt = shared_path->begin();
+    auto rightOpIt = std::prev(std::prev(shared_path->rend()));
+    while (leftOpIt != std::prev(shared_path->end())) {
+        QL_ASSERT(*std::next(leftOpIt) == *rightOpIt);
+        QL_ASSERT(platform->topology->get_distance(*leftOpIt, *rightOpIt) == 1);
```

Comment:
Distance between leftOpIt and rightOpIt needs to be 1, because it is the hop in the topology where the 2q gate will be executed.
What this loop does is go through the proposed path, and create alters for every hop it goes through.

Reply:
#479